### PR TITLE
lots of cleanup

### PR DESCRIPTION
**Pull request by Iepoev**
mirror of #53 but to development branch instead of master

### CODE REVIEW

**Comment by Iepoev** [comment]
closes #43 and #57

File: bin/mympirun.py [hunk 1]
```diff
@@ -31,92 +31,91 @@
...
+
+    # init generaloption with the various mpirun cli options
+    optionparser = MympirunOption(ismpirun=isfake)
+
+    # see if an mpi flavor was explicitly chosen as an argument
+    # if not, just use the mpirun that was called
+    # We are using sys.argv because generaloption depends on the the returned
+    # scriptname
+    setmpi = optionparser.options.setmpi if optionparser.options.setmpi else sys.argv[0]
+    optionparser.log.debug("mympirun will use %s as MPI flavor", setmpi)
```
Comment (stdweird) [inline bin/mympirun.py]: message makes no sense when `sys,argv[0]` is used. don't use teh if/else onliner, and add this message only in the if block

File: bin/mympirun.py [hunk 2]
```diff
@@ -31,92 +31,91 @@
...
+        optionparser.log.raiseException("No sched class found (options.setsched %s ; found Sched classes %s)",
+                                        optionparser.options.setsched, ", ".join(found_sched_names))
     else:
-        mo.log.debug("Found sched class %s from options.schedtype %s (all Sched found %s)" %
-                     (sched.__name__, mo.options.schedtype, ", ".join(found_sched_names)))
+        optionparser.log.debug("Found sched class %s from options.setsched %s (all Sched found %s)",
+                               sched.__name__, optionparser.options.setsched, ", ".join(found_sched_names))
 
-    return mpi, sched, mo
+    if optionparser.args is None or len(optionparser.args) == 0:
```
Comment (stdweird) [inline bin/mympirun.py]: just `if not optionparser.args:`?

File: lib/vsc/mympirun/mpi/intelmpi.py [hunk 3]
```diff
@@ -158,28 +161,26 @@ def mpiexec_set_global_options(self):
             # this only affects libiomp5 usage (ie intel compilers!)
             self.mpiexec_global_options["KMP_AFFINITY"] = "compact"
 
-            """
-            if self.options.hybrid == 4:
-                self.mpiexec_global_options["I_MPI_PIN_DOMAIN"]="cache"
-            elif self.options.hybrid == 2:
-                self.mpiexec_global_options["I_MPI_PIN_DOMAIN"]="socket"
-            """
+            # if self.options.hybrid == 4:
```
Comment (stdweird) [inline lib/vsc/mympirun/mpi/intelmpi.py]: remove if unused

File: lib/vsc/mympirun/mpi/intelmpi.py [hunk 4]
```diff
@@ -24,54 +24,59 @@
...
     _mpiscriptname_for = ['impirun']
     _mpirun_for = ['impi']
     _mpirun_version = lambda x: LooseVersion(x) < LooseVersion("4.1.0.0")
     _mpirun_version = staticmethod(_mpirun_version)
 
-    RUNTIMEOPTION = {'options':{'mpdbulletproof':("Start MPD in bulletproof", None, "store_true", False),
-                                'fallback':("Enable device fallback", None, "store_true", False),
-                                'daplud':("Enable DAPL UD connections", None, "store_true", False),
-                                'xrc':("Enable Mellanox XRC", None, "store_true", False),
+    RUNTIMEOPTION = {'options': {'mpdbulletproof': ("Start MPD in bulletproof", None, "store_true", False),
```
Comment (stdweird) [inline lib/vsc/mympirun/mpi/intelmpi.py]: fix indentation

File: lib/vsc/mympirun/mpi/mpi.py [hunk 5]
```diff
@@ -26,114 +26,139 @@
...
+    # import all modules in this dir: http://stackoverflow.com/a/16853487
+    for loader, modulename, _ in pkgutil.walk_packages([os.path.dirname(__file__)]):
+        loader.find_module(modulename).load_module(modulename)
 
-    # check on scriptname
-    for mpi in found_mpi:
-        if mpi._is_mpiscriptname_for(scriptname):
-            stripfake()  # mandatory before return at this point
-            return scriptname, mpi, found_mpi
+    supp_mpi_impl = get_subclasses(MPI)  # supported MPI implementations, these were imported in __init__.py
```
Comment (stdweird) [inline lib/vsc/mympirun/mpi/mpi.py]: fix comment

File: lib/vsc/mympirun/mpi/mpi.py [hunk 6]
```diff
@@ -210,56 +233,57 @@ def __init__(self, options, cmdargs, **kwargs):
         super(MPI, self).__init__(**kwargs)
 
         # sanity checks
-        if getattr(self, 'id', None) is None:
+        if getattr(self, 'sched_id', None) is None:
             self.log.raiseException("__init__: id None (should be set by one of the Sched classes)")
```
Comment (stdweird) [inline lib/vsc/mympirun/mpi/mpi.py]: `id` - > `sched_id`

File: lib/vsc/mympirun/mpi/mpi.py [hunk 7]
```diff
@@ -210,56 +233,57 @@ def __init__(self, options, cmdargs, **kwargs):
...
 
         if self.cmdargs is None or len(self.cmdargs) == 0:
```
Comment (stdweird) [inline lib/vsc/mympirun/mpi/mpi.py]: `if not self.cmdargs`

File: lib/vsc/mympirun/mpi/mpi.py [hunk 8]
```diff
@@ -268,327 +292,167 @@ def main(self):
...
     def check_usable_cpus(self):
-        """
-        Check and act on fact of non-standard cpus (eg due to cpusets)
-        - default: do nothing more then log
-        """
+        """Check and log if non-standard cpus (eg due to cpusets)."""
         if not self.foundppn == len(self.cpus):
             self.log.info(("check_usable_cpus: non-standard cpus found: requested ppn %s, found cpus %s, "
-                           "usable cpus %s") % (self.ppn, self.foundppn, len(self.cpus)))
+                           "usable cpus %s"), self.ppn, self.foundppn, len(self.cpus))
```
Comment (stdweird) [inline lib/vsc/mympirun/mpi/mpi.py]: move last part of messag eto previous line, get rid of `()` around the message.

File: lib/vsc/mympirun/mpi/mpi.py [hunk 9]
```diff
@@ -268,327 +292,167 @@ def main(self):
...
+        Sets ompthreads to the amount of threads every MPI process should use.
+
+        For example, with hybrid 2 every MPI process should have a total 2 threads (each on a seperate processors).
+        This way each node will have 8 MPI processes (assuming ppn is 16). Will default to 1 if hybrid is disabled.
+        """
         if 'OMP_NUM_THREADS' in os.environ:
-            t = os.environ['OMP_NUM_THREADS']
+            threads = os.environ['OMP_NUM_THREADS']
         else:
             if self.options.hybrid is None or self.options.hybrid == 0:
```
Comment (stdweird) [inline lib/vsc/mympirun/mpi/mpi.py]: `if not self.options.hybrid`

File: lib/vsc/mympirun/mpi/mpi.py [hunk 10]
```diff
@@ -268,327 +292,167 @@ def main(self):
...
+            self.select_device()
 
-        vsmpctl = "vsmpctl --status"
-        ec, out = run_simple_noworries(vsmpctl)
-        if ec > 0:
-            self.log.debug("scalemp_vsmp: vSMP not found (cmd %s ec %s output %s)" % (vsmpctl, ec, out))
-            return
+        device_ip_reg_map = {
+            'eth': r"ether.*?\n.*?inet\s+(\d+\.\d+.\d+.\d+/\d+)",
+            'ib': r"infiniband.*?\n.*?inet\s+(\d+\.\d+.\d+.\d+/\d+)"
```
Comment (stdweird) [inline lib/vsc/mympirun/mpi/mpi.py]: comma after last entry (for future diffs)

File: lib/vsc/mympirun/mpi/mpi.py [hunk 11]
```diff
@@ -268,327 +292,167 @@ def main(self):
...
+        for ipaddr_mask in reg.finditer(out):
+            ip_info = IP(ipaddr_mask.group(1), make_net=True)
+            network_netmask = "%s/%s" % (ip_info.net(), ip_info.netmask())
+            res.append(network_netmask)
+            self.log.debug("set_netmask: convert ipaddr_mask %s into network_netmask %s",
+                           ipaddr_mask.group(1), network_netmask)
 
-        self.options.scalemp_vsmp = True
+        self.log.debug("set_netmask: return complete netmask %s", res)
+        if len(res) > 0:
```
Comment (stdweird) [inline lib/vsc/mympirun/mpi/mpi.py]: `if res:`

File: lib/vsc/mympirun/mpi/mpi.py [hunk 12]
```diff
@@ -268,327 +292,167 @@ def main(self):
...
-
-    def get_universe_ncpus(self):
-        """Return ppn for universe"""
-        return self.mpitotalppn
+    def set_device(self, founddev):
+        """Set self.device to founddev, but doublecheck if the path to this device actually exists """
+        self.device = self.DEVICE_MPIDEVICE_MAP[founddev]
+        path = self.DEVICE_LOCATION_MAP[founddev]
+        if path is None or os.path.exists(path):
+            self.log.warning("Forcing device %s (founddevice %s), but path %s not found.",
```
Comment (stdweird) [inline lib/vsc/mympirun/mpi/mpi.py]: but test above is `or  os.path.exists(path)`. should be `or not  os.path.exists(path)`?

File: lib/vsc/mympirun/mpi/mpi.py [hunk 13]
```diff
@@ -643,240 +531,155 @@ def set_pinning(self, mp=None):
...
         self.make_mpdboot_options()
 
-        self.log.debug("make_mpdboot set options %s" % self.mpdboot_options)
+        self.log.debug("make_mpdboot set options %s", self.mpdboot_options)
+
+    def mpdboot_set_localhost_interface(self):
+        """Sets mpdboot_localhost_interface to the first result of get_localhosts()."""
+        localhosts = self.get_localhosts()
+        if len(localhosts) > 1:
+            self.log.warning(("set_mpd_localhost_interface: more then one match for localhost from unique nodes "
```
Comment (stdweird) [inline lib/vsc/mympirun/mpi/mpi.py]: `set_mpd_localhost_interface` -> `mpd_set_localhost_interface`

File: lib/vsc/mympirun/mpi/mpi.py [hunk 14]
```diff
@@ -643,240 +531,155 @@ def set_pinning(self, mp=None):
...
+                                   idx, iface, nodename, ip)
+
+                    res.append((nodename, iface))
+                else:
+                    self.log.debug("get_localhost idx %s: no interface match for prefixes %s out %s",
+                                   idx, iface_prefix, out)
+            else:
+                self.log.error("get_localhost idx %s: cmd %s failed with output %s", idx, cmd, out)
+
+        if len(res) == 0:
```
Comment (stdweird) [inline lib/vsc/mympirun/mpi/mpi.py]: `if not res`

File: lib/vsc/mympirun/mpi/mpi.py [hunk 15]
```diff
@@ -885,45 +688,42 @@ def make_mpiexec_options(self):
...
             else:
-                self.log.debug("make_mpiexe_hydra_options: no rmk from HYDRA_RMK %s and hydra_info %s" %
-                               (self.HYDRA_RMK, self.hydra_info))
+                self.log.debug("make_mpiexe_hydra_options: no rmk from HYDRA_RMK %s and hydra_info %s",
+                               self.HYDRA_RMK, self.hydra_info)
         else:
             launcher = None
             if getattr(self, 'HYDRA_LAUNCHER', None) is not None:
                 launcher = [x for x in self.HYDRA_LAUNCHER if x in self.hydra_info.get('launcher', [])]
                 if len(launcher) > 0:
```
Comment (stdweird) [inline lib/vsc/mympirun/mpi/mpi.py]: `if launcher:`

File: lib/vsc/mympirun/mpi/mpi.py [hunk 16]
```diff
@@ -932,38 +732,38 @@ def make_mpiexec_hydra_options(self):
                 self.mpiexec_options.append("-%s %s" % (self.HYDRA_LAUNCHER_NAME, launcher[0]))
 
             if launcher_exec is not None:
-                self.log.debug("make_mpiexec_hydra_options: HYDRA using launcher exec %s" % launcher_exec)
+                self.log.debug("make_mpiexec_hydra_options: HYDRA using launcher exec %s", launcher_exec)
                 self.mpiexec_options.append("-%s-exec %s" % (self.HYDRA_LAUNCHER_NAME, launcher_exec))
 
     def get_hydra_info(self):
-        """Get a dict with hydra info"""
+        """Get a dict with hydra info."""
```
Comment (stdweird) [inline lib/vsc/mympirun/mpi/mpi.py]: i find this very odd that this data came from `-info` but now works with `-help`. please double check

File: lib/vsc/mympirun/mpi/mpi.py [hunk 17]
```diff
@@ -932,38 +732,38 @@ def make_mpiexec_hydra_options(self):
...
             if key is None:
-                self.log.raiseException("get_hydra_info: failed to get hydra info: missing key in %s (out: %s)" %
-                                        (r.groupdict(), out))
+                self.log.raiseException("get_hydra_info: failed to get hydra info: missing key in %s (out: %s)",
+                                        regex.groupdict(), out)
             key = key.strip().lower()
-            value = r.groupdict()['value']
+            value = regex.groupdict()['value']
             if value is None:
                 self.log.debug("get_hydra_info: failed to get hydra info: missing value in %s (out: %s)" %
```
Comment (stdweird) [inline lib/vsc/mympirun/mpi/mpi.py]: no templating, use `,`

File: lib/vsc/mympirun/mpi/mpi.py [hunk 18]
```diff
@@ -932,38 +732,38 @@ def make_mpiexec_hydra_options(self):
...
-                               (r.groupdict(), out))
+                               (regex.groupdict(), out))
                 value = ''
             values = [x.strip().strip('"').strip("'") for x in value.split() if len(x.strip()) > 0]
```
Comment (stdweird) [inline lib/vsc/mympirun/mpi/mpi.py]: `if x.strip()`

File: lib/vsc/mympirun/mpi/mpi.py [hunk 19]
```diff
@@ -932,38 +732,38 @@ def make_mpiexec_hydra_options(self):
...
             hydra_info[key] = values
-        self.log.debug("get_hydra_info: found info %s" % hydra_info)
+        self.log.debug("get_hydra_info: found info %s", hydra_info)
 
         keymap = {"rmk": r'^resource\s+management\s+kernel.*available',
```
Comment (stdweird) [inline lib/vsc/mympirun/mpi/mpi.py]: fix indentation

File: lib/vsc/mympirun/rm/sched.py [hunk 20]
```diff
@@ -25,40 +25,60 @@
...
+    for loader, modulename, _ in pkgutil.walk_packages([os.path.dirname(__file__)]):
+        loader.find_module(modulename).load_module(modulename)
 
-def whatSched(requested):
-    """Return the scheduler class"""
     found_sched = get_subclasses(Sched)
+
+    # if there is no requested scheduler set or if the script is not in a job
+    # environment, default to local
+    if "SCHED_ENVIRON_ID" not in os.environ or requested is None:
```
Comment (stdweird) [inline lib/vsc/mympirun/rm/sched.py]: should be `found_sched.SCHED_ENVIRON_ID not in os.environ ...`

File: lib/vsc/mympirun/rm/sched.py [hunk 21]
```diff
@@ -25,40 +25,60 @@
...
+        for sched in found_sched:
+            if sched._is_sched_for("local"):
+                return sched, found_sched
```
Comment (stdweird) [inline lib/vsc/mympirun/rm/sched.py]: raiseException for the else block? at least log warning

File: lib/vsc/mympirun/rm/sched.py [hunk 22]
```diff
@@ -121,97 +141,82 @@ def _is_sched_for(cls, name=None):
...
         """
         if self.foundppn is None:
             self._cores_on_this_node()
 
         try:
-            cs = sched_getaffinity()  # get affinity for current proc
-            self.cpus = [idx for idx, cpu in enumerate(cs.cpus) if cpu == 1]
-        except:
+            proc_affinity = sched_getaffinity()  # get affinity for current proc
+            self.cpus = [idx for idx, cpu in enumerate(proc_affinity.cpus) if cpu == 1]
```
Comment (stdweird) [inline lib/vsc/mympirun/rm/sched.py]: add debug log in both try and except block

File: setup.py [hunk 23]
```diff
@@ -84,16 +84,16 @@ def run(self):
                     os.symlink(rel_script, sym_name)
                     newoutfile = os.path.join(rel_script_dir, sym_name)
                     self.outfiles.append(newoutfile)
-                    log.info("symlink %s to %s newoutfile %s" % (rel_script, sym_name, newoutfile))
+                    log.info("symlink %s to %s newoutfile %s", rel_script, sym_name, newoutfile)
```
Comment (stdweird) [inline setup.py]: this is not fancylogger or logging module. make sure you have tested this (e.g. look at the  messages when doing a `python setup.py install --user`)

**Comment by Iepoev** [comment]
should be done now

File: lib/vsc/mympirun/rm/sched.py [hunk 24]
```diff
@@ -47,14 +47,22 @@ def what_sched(requested):
...
+
+    # first, try to use the scheduler that was requested
+    if requested:
         for sched in found_sched:
-            if sched._is_sched_for("local"):
+            if sched._is_sched_for(requested):
                 return sched, found_sched
 
-    # search for a scheduler that matches requested
+    # next, try to use the scheduler defined by environment variables
```
Comment (stdweird) [inline lib/vsc/mympirun/rm/sched.py]: this is not next, this is `else`

File: setup.py (hunk 23 above)
Comment (Iepoev) [inline setup.py]: I checked it and it logs to stdout. should it do more than this?

File: lib/vsc/mympirun/rm/sched.py [hunk 25]
```diff
@@ -45,26 +46,25 @@ def what_sched(requested):
...
 
-    # if there is no requested scheduler set or if the script is not in a job
-    # environment, default to local
-
     # first, try to use the scheduler that was requested
     if requested:
         for sched in found_sched:
             if sched._is_sched_for(requested):
                 return sched, found_sched
+        LOGGER.warn("%s scheduler was requested, but mympirun failed to find an implementation" % requested)
```
Comment (stdweird) [inline lib/vsc/mympirun/rm/sched.py]: no templating

File: lib/vsc/mympirun/rm/sched.py [hunk 26]
```diff
@@ -45,26 +46,25 @@ def what_sched(requested):
...
-        for sched in found_sched:
-            if sched.SCHED_ENVIRON_ID in os.environ:
-                return sched, found_sched
+    for sched in found_sched:
+        if sched.SCHED_ENVIRON_ID in os.environ:
+            return sched, found_sched
 
     # If that fails, try to force the local scheduler
     for sched in found_sched:
+        LOGGER.debug("No scheduler found in environment, defaulting to local")
```
Comment (stdweird) [inline lib/vsc/mympirun/rm/sched.py]: `defaulting` -> `trying`

File: setup.py (hunk 23 above)
Comment (stdweird) [inline setup.py]: no, main concern was to check if the templating was done

**Comment by stdweird** [comment]
@Iepoev thanks for this round of cleanup. please rebase your other branches on develop branch

**Pull request merged**